cogs/boolean_utils.py

```diff
@@ -0,0 +1,395 @@
+    # Defines operator precedence
+    prec_dict = {
+        '!' : 0, # NOT
+        '*' : 1, # AND
+        '^' : 2, # XOR
+        '+' : 3, # OR
+        '(' : 8,
+        ')' : 9,
+        -1  : 10 # op_stack is empty, skip
+    }
+
+    def __init__(self, exp):
+        '''Initialize all values'''
+        self.var_set.clear()
+        self.disp_exp = exp
+        self.orig_exp = "".join(exp.split())
+        self.op_stack[:] = []
+        self.var_stack[:] = []
+        self.error = False
+        self.error_msg = ""
+        # Create post-fix expression
+        self.process_exp(self.orig_exp)
+
+    ######################################################################
+    ####### Parse in-fix expression into post-fix ########################
+    ######################################################################
+
+    def process_exp(self, exp):
+        """
+        Processes an in-fix expression and saves a post-fix version
+        of the expression on the class level. The post-fix expression
+        is easier to evaluate in code and makes creating the truth
+        table more efficient.
+        """
+        self.post_exp = ""
+        # Step through chars of booelean expression
+        for char in exp:
+            if self.error == True:
+                return
+            if char == ' ':
+                continue
+            # Alphas are variables
+            elif char.isalpha():
+                self.var_set.add(char)
+                self.var_stack.append(char)
+                self.post_exp += char
+            # Start parens
+            elif char == '(':
+                self.op_stack.append(char)
+            # End parens
+            elif char == ')':
+                # Process all operators until start paren
+                while self.get_top(self.op_stack) != '(':
+                    self.process_an_op()
+                    if self.get_top(self.op_stack) == -1:
+                        self.error = True
+                        self.error_msg = "Unbalanced paren: )"
+                        return
+                # Pop start paren
+                if self.op_stack:
+                    self.op_stack.pop()
+            # If the char is a valid operator then process it
+            elif self.is_valid_op(char):
+                self.process_char(char)
+            else:
+                # we have found an invalid char
+                self.error = True
+                self.error_msg = "{} is not a valid symbol!".format(char)
+        if self.error == True:
+            return
+        # Process remaining operators
+        while self.op_stack and len(self.var_stack) > 0:
+            if self.get_top(self.op_stack) == '(':
+                self.error = True
+                self.error_msg = "Unbalanced paren: ("
+            self.process_an_op()
+
+    def is_valid_op(self, char):
+        '''True if char is a valid boolean operator'''
+        result = char == '+' or char == '^' or char == '*' or char == '!'
```

Comment:
Should we make these class constants as a list and then do something like `return char in boolean_operators`

Reply:
That would be a much more pythonic way to handle that